### PR TITLE
modernize and clean up mayaUsd and mayaUsdPlugin targets!

### PR DESCRIPTION
**Pull request by HamedSabri-adsk**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,10 +12,9 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-# =======================================================================
 #
 
-cmake_minimum_required(VERSION 3.12.0)
+cmake_minimum_required(VERSION 3.13...3.17)
```
Comment (HamedSabri-adsk) [inline CMakeLists.txt]: cmake_policy(VERSION <min>[...<max>]).  This syntax was introduced in CMake V12.4.

I've set the min policy to 3.13, although I think 3.15 is more suitable.

File: cmake/utils.cmake [hunk 2]
```diff
@@ -148,18 +148,37 @@ function(mayaUsd_install_rpath rpathRef NAME)
 endfunction()
 
 function(mayaUsd_promoteMayaUsdHeader)
-    set(srcFile ${CMAKE_CURRENT_SOURCE_DIR}/base/mayaUsd.h.src)
+    set(srcFile ${CMAKE_CURRENT_SOURCE_DIR}/mayaUsd.h.src)
```
Comment (HamedSabri-adsk) [inline cmake/utils.cmake]: Get rid off hard-coded `base` directory.

File: cmake/utils.cmake [hunk 3]
```diff
@@ -203,7 +222,7 @@ function(mayaUsd_copyFiles target)
         message(FATAL_ERROR "DESTINATION keyword is not specified.")
     endif()
```
Comment (HamedSabri-adsk) [inline cmake/utils.cmake]: mayaUsd_promoteHeaderList function now takes a TARGET, DIRNAME, and HEADERS which are needed to be passed explicitly.

File: lib/CMakeLists.txt [hunk 4]
```diff
@@ -311,134 +132,22 @@ install(TARGETS ${LIBRARY_NAME}
...
-            )
-        endforeach()
-    endif()
-
-endif()
-
-if(IS_MACOSX)
-    set(_macDef OSMac_)
+add_subdirectory(ufe)
 endif()
```
Comment (HamedSabri-adsk) [inline lib/CMakeLists.txt]: Cleaned up the top level CMake. This file was getting really large and was hard to follow up what is happening there.

All the sub-directories (base, listeners, nodes, utils,render,ufe) now have their own CMakeLists.  As of CMakeV12.4, 
The target_* directive calls don't need to be part of the same CMakeLists.txt where add_executable or add_library directives are called.

File: lib/base/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# -----------------------------------------------------------------------------
+# sources
+# -----------------------------------------------------------------------------
+target_sources(${TARGET_NAME} 
+    PRIVATE
+        debugCodes.cpp
+)
```
Comment (HamedSabri-adsk) [inline lib/base/CMakeLists.txt]: Always use target_sources!

File: cmake/utils.cmake [hunk 6]
```diff
@@ -148,18 +148,37 @@ function(mayaUsd_install_rpath rpathRef NAME)
...
         message(STATUS "promoting: " ${srcFile})
     endif()
     configure_file(${srcFile} ${dstFile})
 endfunction()
 
-function(mayaUsd_promoteHeaderList)
-    foreach(header ${ARGV})
+function(mayaUsd_promoteHeaderList target)
+    cmake_parse_arguments(PREFIX
+        "TARGET"
```
Comment (seando-adsk) [inline cmake/utils.cmake]: From the doc, the first arg is the options. "target" is already a arg passed to the function. I had already made this change in my local branch like so:
```cmake
    cmake_parse_arguments(PREFIX 
        ""          # options
        "SUBDIR"    # one_value keywords
        "FILES"     # multi_value keywords
        ${ARGN}
    )
```
Why is DIRNAME mandatory? 

I had also fixed the copyDirectory below.

File: lib/ufe/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,207 @@
+# -----------------------------------------------------------------------------
```
Comment (seando-adsk) [inline lib/ufe/CMakeLists.txt]: I'm assuming that this block was just copied from the other cmake file and nothing has really changed in it. If something did change can you please point it out?

File: plugin/adsk/plugin/CMakeLists.txt [hunk 8]
```diff
@@ -1,50 +1,99 @@
...
+target_compile_definitions(${TARGET_NAME}
     PRIVATE
         MAYAUSD_PLUGIN_EXPORT
 )
 
-target_include_directories(${PLUGIN_PACKAGE}
+if(UFE_FOUND)
+    target_compile_definitions(${TARGET_NAME} 
+        PRIVATE
+            -DWANT_UFE_BUILD
```
Comment (seando-adsk) [inline plugin/adsk/plugin/CMakeLists.txt]: I'm thinking that rather than have this in a separate if, it could be done using a "generator expression". Something like this just after line 50:

```cmake
$<UFE_FOUND:WANT_UFE_BUILD>
```

Note: you also don't need the -D as it will be removed (as per the docs).

Note: You can also use the same syntax for the target_include_directories below.

Comment (HamedSabri-adsk) [reply plugin/adsk/plugin/CMakeLists.txt]: @seando-adsk  Cool! Thanks for pointing it out. Even better! will change it.

**Comment by ppt-adsk** [comment]
Hey Hamed, are you sure this doesn't conflict with
https://github.com/Autodesk/maya-usd/tree/Autodesk/ppt/refactoring_sandbox_to_dev_00
?  See
https://github.com/Autodesk/maya-usd/pull/154
I'm wondering about lib/CMakeLists.txt, for example.

File: lib/ufe/CMakeLists.txt (hunk 7 above)
Comment (HamedSabri-adsk) [reply lib/ufe/CMakeLists.txt]: Correct. Nothing really changed here, expect the sub-directory prefix that is gone now:

https://github.com/Autodesk/maya-usd/blob/334926c3a6e965fcea048fb76f3980cf50e3f002/lib/CMakeLists.txt#L80

File: cmake/utils.cmake (hunk 6 above)
Comment (HamedSabri-adsk) [reply cmake/utils.cmake]: @seando-adsk 
Right, I didn't really have to pass the target and was solely added for debugging purposes.  Target is not really used inside the function.

Both DIRNAME  and HEADERS keywords are mandatory. DIRNAME is needed because we no longer have the header files with the subdir prefix ( e.g `base/api.h`, `base/debugCodes.h`, etc... ). Instead the subdir name now needs to be passed to mayaUsd_promoteHeaderList function as a separate argument.

```
e.g
set(headers
    api.h
    debugCodes.h
)

I am going to change get rid of the target and rename DIRNAME to SUBDIR. 
e.g
mayaUsd_promoteHeaderList(HEADERS ${headers} SUBDIR base)
mayaUsd_promoteHeaderList(HEADERS ${headers} SUBDIR nodes)
mayaUsd_promoteHeaderList(HEADERS ${headers} SUBDIR listeners)
```

Did you remove target from mayaUsd_copyDirectory? I believe target was needed to be passed to `add_custom_command` command?

Comment (seando-adsk) [reply cmake/utils.cmake]: @HamedSabri-adsk Okay I see for both those keywords being mandatory now. No I didn't remove TARGET from any of the functions.

**Comment by HamedSabri-adsk** [comment]
> 
> 
> Hey Hamed, are you sure this doesn't conflict with
> https://github.com/Autodesk/maya-usd/tree/Autodesk/ppt/refactoring_sandbox_to_dev_00
> ? See
> #154
> I'm wondering about lib/CMakeLists.txt, for example.

Hey Pierre, yes indeed. I was aware of this and wanted to discuss with you. 
I was thinking to wait for #154 to be merged into dev first. Then I would rebase my branch in this PR (obviously I need to add more new changes) and then have this branch merge to dev. 

What do you think?

**Comment by kxl-adsk** [comment]
@HamedSabri-adsk there are couple of other PRs which will wait for #154. This one should wait as well.

File: plugin/adsk/plugin/CMakeLists.txt [hunk 9]
```diff
@@ -50,12 +50,10 @@ target_compile_definitions(${TARGET_NAME}
...
 
-if(UFE_FOUND)
-    target_compile_definitions(${TARGET_NAME} 
-        PRIVATE
-            -DWANT_UFE_BUILD
-    )
-endif()
+target_compile_definitions(${TARGET_NAME} 
+    PRIVATE
+        $<$<BOOL:${UFE_FOUND}>:WANT_UFE_BUILD>
```
Comment (HamedSabri-adsk) [inline plugin/adsk/plugin/CMakeLists.txt]: @seando-adsk I used generator expression $<BOOL:...> to convert FALSE or TRUE to 0 or 1

`$<UFE_FOUND:WANT_UFE_BUILD> --> this would cause an error because ${UFE_FOUND} isn't 1 or 0`

`$<$<BOOL:${UFE_FOUND}>:WANT_UFE_BUILD>   --> BOOL:${UFE_FOUND} converts TRUE or FALSE to 1 or 0`

File: cmake/utils.cmake [hunk 10]
```diff
@@ -178,7 +178,7 @@ function(mayaUsd_promoteHeaderList target)
 
     foreach(header ${headerFiles})
         set(srcFile ${CMAKE_CURRENT_SOURCE_DIR}/${header})
-        set(dstFile ${CMAKE_BINARY_DIR}/include/mayaUsd/${dirName}/${header})
+        set(dstFile ${CMAKE_BINARY_DIR}/include/mayaUsd/${subDir}/${header})
 
         set(content "#pragma once\n#include \"${srcFile}\"\n")
```
Comment (HamedSabri-adsk) [inline cmake/utils.cmake]: @seando-adsk  target is gone and I renamed DIRNAME to SUBDIR. Thanks.

File: plugin/adsk/plugin/CMakeLists.txt (hunk 9 above)
Comment (seando-adsk) [reply plugin/adsk/plugin/CMakeLists.txt]: So the whole point was to merge this into the one above at line 48. Just move line 55 to just after 50. Then you don't need two target_compile_definitions. Same for the target_include_directories below.

Comment (HamedSabri-adsk) [reply plugin/adsk/plugin/CMakeLists.txt]: Ugh d’accord ! je suis aveugle...

Comment (seando-adsk) [reply plugin/adsk/plugin/CMakeLists.txt]: Excellent thanks.

**Comment by HamedSabri-adsk** [comment]
Closing this PR. I am internally working on this since there is going to be more new changes coming in. Will create a separate PR once it's ready.
